### ISSUE TITLE
Subissue 3: Implement solver classes with backend registry equivalent to current CMF methods

### ISSUE THREAD

**Issue by adku1173** [acoular/acoular#630]
> *This was generated by AI during triage.*

# Subissue 3: Implement solver classes with backend registry equivalent to current CMF methods

**Issue type:** Feature request / refactoring subissue  
**Labels:** `enhancement`, `new feature`  
**Type:** AFK  
**Blocked by:** #628, #629  
**Parent:** #632

## Feature Description

Implement concrete solver classes that reproduce the behavior of the solver methods currently called by `BeamformerCMF`, but do not yet replace the current `BeamformerCMF._calc` solver dispatch. Solver classes should be named by their solver family/objective contract and should select package implementations through a `backend` trait and backend registry. They should not be CMF-specific classes.

After Subissue 2, the abstract problem/solver interfaces, `SolverOutput`, problem scaling registry, and problem-level `unit_mult` handling should exist. This subissue should add concrete solver classes plus a backend registry for the same algorithms/objective variants that CMF currently invokes internally. The tests should verify that solving the same assembled CMF problem through the new problem/solver path gives the same result as the current CMF implementation.

This is a preparation step. `BeamformerCMF._calc` should still use its current solver branches after this subissue. The actual replacement of those branches with the new interface belongs to Subissue 4. If needed, this subissue may be narrowed or split further; it is not required to force all wrappers into one change if the interface needs more refinement.

## Use Case

Before changing `BeamformerCMF._calc`, we need confidence that the new solver classes are numerically equivalent to the existing implementation. This subissue provides that safety net by implementing the equivalent solvers and comparing their output against the current CMF behavior.

For users, this should not change CMF behavior yet. For maintainers, it creates tested solver building blocks that can be wired into CMF in the next subissue.

## What to build

Add concrete solver classes for the solver methods currently used by `BeamformerCMF`. Solver classes should expose a `backend` trait and dispatch through a backend registry:

- `LassoLarsSolver(backend='sklearn')` for current `LassoLars` behavior,
- `LassoLarsBICSolver(backend='sklearn')` for current `LassoLarsBIC` behavior,
- `OMPCVSolver(backend='sklearn')` for current `OMPCV` behavior,
- `NNLSSolver(backend='sklearn')` for current CMF `NNLS` behavior,
- `LBFGSBSolver(backend='scipy')` for current `fmin_l_bfgs_b` behavior,
- `FISTALassoSolver(backend='pylops')` for current PyLops `FISTA` LASSO behavior,
- `SplitBregmanLassoSolver(backend='pylops')` for current PyLops `Split_Bregman` LASSO behavior.

The concrete classes should use the problem/solver interfaces from Subissue 2 and return `SolverOutput` objects. Public classes should be organized by stable solver family/objective contract, not by package name alone. For example, prefer `NNLSSolver(backend='sklearn')` over `SklearnNNLS`, and prefer `FISTALassoSolver(backend='pylops')` over a vague `FISTASolver` when the LASSO objective determines which parameters and proximal behavior are active. Avoid CMF-specific names such as `CmfLassoLarsSolver` unless an implementation is genuinely CMF-specific.

The implementation should preserve the current CMF numerical conventions:

- column-wise dictionary scaling where currently used, owned by `LeastSquaresProblem.dictionary_scaling`,
- problem-level `unit_mult` working-unit scaling and inverse scaling, owned by `LeastSquaresProblem`,
- non-negative solution constraints where current methods require them,
- `alpha`, `n_iter`, and `show` semantics,
- PyLops availability checks for `FISTA` and `Split_Bregman`.

`unit_mult` must not be implemented as a generic solver trait or as a registered problem scaling transform. Solver classes receive the problem instance and may inspect `problem.unit_mult` only where needed to preserve legacy backend parameter conventions. For example, current sklearn Lasso-style branches use `alpha * unit_mult`, while the current PyLops CMF branches use `alpha` unchanged.

Tests should assemble the same CMF problem data and compare:

1. the result produced by the current `BeamformerCMF(method=...)` implementation, and
2. `SolverOutput.solution` produced by the new problem/solver path for the same method, dictionary matrix, vectorized CSM data, start value, problem-level `unit_mult`, and relevant solver parameters.

The equivalence setup should let `LeastSquaresProblem` apply dictionary/data scaling and `unit_mult`; tests should not duplicate that scaling manually unless they are explicitly testing the legacy inline reference path.

The current `BeamformerCMF._calc` solver dispatch should remain unchanged in this subissue so that it can serve as the reference implementation for these equivalence tests.

## Backend registry expectations

Introduce a solver backend registry separate from the problem scaling registry. The registry should be keyed by solver family/objective variant and backend name, for example:

```python
register_solver_backend('nnls', 'sklearn', solve_func, dependency='sklearn')
register_solver_backend('fista_lasso', 'pylops', solve_func, dependency='pylops')
```

Expected behavior:

- `backend` is a string validated against registered backend names for that solver family/objective variant.
- Invalid backend names fail on trait assignment or equivalent validation, not only at solve time.
- Optional but unavailable backends may still be registered. Constructing/configuring a solver with such a backend is allowed. Calling `solve(...)` raises a clear `ImportError` or dependency-specific error.
- Distinguish registered backends from currently available backends. For example, if PyLops is not installed, `registered_solver_backends('fista_lasso')` may include `'pylops'`, while `available_solver_backends('fista_lasso')` should not.
- Provide an inspection helper that exposes backend metadata, including availability and optional dependency name.
- Backend solve functions receive the Acoular context:

```python
solve_func(solver, problem, dictionary_matrix, data, start_value)
```

This lets a backend read solver parameters such as `n_iter` or `show`, problem parameters such as `alpha`, `nonnegative`, or `unit_mult`, and return backend diagnostics through `SolverOutput.info`.

A backend change is a real numerical choice. Backends under one solver class promise the same solver family/objective contract, but not bitwise-identical behavior. Legacy equivalence tests should therefore compare the default backend selected to reproduce current CMF behavior.

## Backend-specific keyword arguments

Solver classes may expose `backend_kwargs` for options that belong only to a selected backend and are not stable solver-family traits. Acoular should not duplicate backend function signatures in the registry. Instead, `backend_kwargs` are passed through to the selected backend adapter or underlying package call.

Example:

```python
problem = LeastSquaresProblem(
    nonnegative=True,
    solver=NNLSSolver(
        backend='scipy',
        backend_kwargs={
            'maxiter': 1000,
            'atol': 1e-12,
        },
    ),
)
```

A SciPy backend adapter may then forward those kwargs directly:

```python
def scipy_nnls_backend(solver, problem, dictionary_matrix, data, start_value=None):
    solution, residual = scipy.optimize.nnls(
        dictionary_matrix,
        data,
        **solver.backend_kwargs,
    )
    return SolverOutput(solution=solution, info={'residual': residual})
```

If a user provides an unsupported keyword, the selected backend implementation should raise the resulting backend error, for example a `TypeError` from SciPy. Backend adapters must not silently ignore `backend_kwargs`. If a backend adapter cannot safely forward keyword arguments, it should raise a clear `ValueError` when `backend_kwargs` is non-empty.

`backend_kwargs` are backend-specific. Changing `backend` may make previously valid `backend_kwargs` invalid. This is acceptable and should fail clearly. `backend_kwargs` must be included in solver digest dependencies because they can affect numerical results.

## Unit multiplier transition expectations

Backward compatibility for existing `unit_mult` behavior is required. The current inline CMF and DAMAS+ implementations multiply the right-hand side data by `unit_mult` and divide the returned solution by `unit_mult`. Some sklearn Lasso-style branches also multiply `alpha` by `unit_mult`.

For this subissue:

- `unit_mult` is considered a problem-level working-unit scale introduced by Subissue 2.
- Generic backend wrappers should not own `unit_mult` as a primary solver trait.
- Generic backend wrappers may inspect `problem.unit_mult` when reproducing legacy backend parameter conventions.
- `unit_mult` remains public on existing beamformers until a later compatibility/deprecation step changes that API.
- `BeamformerCMF._calc` remains unchanged, so current beamformer-level `unit_mult` behavior remains the reference implementation.

## CMF configuration transition expectations

Do not add string-based `BeamformerCMF` problem or solver selector traits in this subissue. The preferred new advanced API is explicit problem composition:

```python
BeamformerCMF(
    problem=L1RegularizedLeastSquaresProblem(
        alpha=...,
        unit_mult=...,
        solver=LassoLarsSolver(backend='sklearn', n_iter=...),
    ),
)
```

Legacy traits such as `method`, `alpha`, `n_iter`, `show`, and `unit_mult` remain for backward compatibility. Later routing work should treat `method` as a deprecated compatibility shortcut that constructs the default problem/solver only when no explicit problem is supplied.

## Code references

Target code on `master` (`9ac7411dc90e7da95d140ca721cce081544d902d`):

- [`BeamformerCMF` configuration traits and current digest dependencies](https://github.com/acoular/acoular/blob/9ac7411dc90e7da95d140ca721cce081544d902d/acoular/fbeamform.py#L1587-L1659)
- [Current inline CMF problem assembly and solver execution in `_calc`](https://github.com/acoular/acoular/blob/9ac7411dc90e7da95d140ca721cce081544d902d/acoular/fbeamform.py#L1661-L1807)
- [Current scikit-learn-backed solver selection](https://github.com/acoular/acoular/blob/9ac7411dc90e7da95d140ca721cce081544d902d/acoular/fbeamform.py#L1715-L1722)
- [Current PyLops-backed `Split_Bregman` and `FISTA` branches](https://github.com/acoular/acoular/blob/9ac7411dc90e7da95d140ca721cce081544d902d/acoular/fbeamform.py#L1724-L1760)
- [Current `fmin_l_bfgs_b` branch and scikit-learn fitting path](https://github.com/acoular/acoular/blob/9ac7411dc90e7da95d140ca721cce081544d902d/acoular/fbeamform.py#L1761-L1807)

WIP guidance from earlier solver experiments:

- [Prototype PyLops least-squares solver implementation](https://github.com/acoular/acoular/blob/3434b8ba45b322d89fe54870368ccf4b8e6ac12d/acoular/solvers/pylops/solver.py#L31-L132)

## Acceptance criteria

- [ ] Concrete solver classes with backend selection exist for the solver methods currently called by `BeamformerCMF`, or an equivalent maintainable class structure covers all existing CMF methods.
- [ ] The new solver classes use the problem/solver interfaces and `SolverOutput` introduced in Subissue 2.
- [ ] `BeamformerCMF._calc` still uses the current solver dispatch and is not yet routed through the new interface.
- [ ] Tests compare each new solver class/default backend against the current `BeamformerCMF(method=...)` result for the same assembled CMF problem.
- [ ] `LassoLars`, `LassoLarsBIC`, `OMPCV`, and `NNLS` match the current CMF results within appropriate floating-point tolerances.
- [ ] `fmin_l_bfgs_b` matches the current CMF result within appropriate floating-point tolerances.
- [ ] `FISTALassoSolver(backend='pylops')` and `SplitBregmanLassoSolver(backend='pylops')` match the current CMF result when PyLops is available.
- [ ] PyLops-backed solver backends remain optional: registered but unavailable when PyLops is missing, and raising a clear import error when used.
- [ ] Problem-level `unit_mult`, `alpha`, `n_iter`, and `show` keep their existing behavior in the new problem/solver path.
- [ ] Solver-specific diagnostic output, if any, is returned through `SolverOutput.info` in a documented/testable way.
- [ ] Public solver classes are named by solver family/objective contract rather than backend package or CMF usage unless genuinely backend- or CMF-specific.
- [ ] Solver classes and backend functions do not duplicate dictionary/data scaling that belongs to `LeastSquaresProblem`.
- [ ] `unit_mult` is not added as a generic solver trait or registered problem scaling transform.
- [ ] Sklearn Lasso-style wrappers preserve the current `alpha * unit_mult` convention where applicable.
- [ ] PyLops wrappers preserve the current CMF convention of using `alpha` unchanged where applicable.
- [ ] At least one equivalence test covers a non-default `unit_mult` value if feasible.
- [ ] Backend registry APIs distinguish registered backends from currently available backends.
- [ ] Invalid backend names fail during backend assignment or equivalent validation.
- [ ] Optional unavailable backends can be configured but fail clearly at solve time.
- [ ] Backend functions receive `(solver, problem, dictionary_matrix, data, start_value)`.
- [ ] `NNLSSolver(backend='sklearn')` reproduces current CMF `NNLS` behavior.
- [ ] No string-based `BeamformerCMF` problem or solver selector traits are added in this subissue.
- [ ] Solver classes expose `backend_kwargs` for backend-specific pass-through options.
- [ ] `backend_kwargs` are included in solver digest dependencies.
- [ ] Backend adapters do not silently ignore non-empty `backend_kwargs`.
- [ ] The backend registry does not duplicate backend signatures via an `accepted_kwargs` mechanism in this subissue.
- [ ] Existing tests pass.
